### PR TITLE
Implict conversions rolling

### PR DESCRIPTION
**Pull request by CursedRock17**
### Description

This pull request is part of many that are needed to update the total repo based on issue #880 following cpp_best_practices of implicit conversions. I targetted all functions with operator conversions and single parameter contstructor that weren't copy based, so initializer lists are included. In order to do this I have been grepping throughout the repository.

### Checklist
- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit2/blob/main/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://moveit.picknik.ai/humble/doc/examples/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit2/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

File: moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h [hunk 1]
```diff
@@ -76,7 +76,7 @@ class JointModelGroup
     }
 
     /// Return a flag indicating whether the state of the solver is initialized
-    operator bool() const
+    explicit operator bool() const
```
Comment (ChrisThrasher) [inline moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: I think implicit conversions like this are useful since you can write `if (joint_mode_group)` instead of `if (bool{joint_model_group})`. Ultimately it's a design decision for you to make whether you want convenience or explicitness, but I think it's no less safe to let these conversions happen quietly.

File: moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp [hunk 2]
```diff
@@ -245,6 +245,14 @@ MoveItCpp::execute(const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
...
   moveit_msgs::msg::RobotTrajectory robot_trajectory_msg;
   robot_trajectory->getRobotTrajectoryMsg(robot_trajectory_msg);
+<<<<<<< HEAD
+<<<<<<< HEAD
+  trajectory_execution_manager_->push(robot_trajectory_msg, controllers);
+  trajectory_execution_manager_->execute();
+  return trajectory_execution_manager_->waitForExecution();
+=======
+=======
+>>>>>>> parent of f117b24a6 (Removed Blocking)
```
Comment (ChrisThrasher) [inline moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp:255]: Merge conflict markers

File: moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp [hunk 3]
```diff
@@ -258,6 +266,10 @@ MoveItCpp::execute(const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
     trajectory_execution_manager_->execute();
     return trajectory_execution_manager_->waitForExecution();
   }
+<<<<<<< HEAD
+>>>>>>> parent of f117b24a6 (Removed Blocking)
+=======
+>>>>>>> parent of f117b24a6 (Removed Blocking)
```
Comment (ChrisThrasher) [inline moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp:272]: Merge conflict markers

**Review by ChrisThrasher — APPROVED** [review]
I agree that `explicit` is generally better than the lack of it but I'm 100% on board with explicit type conversion operators. It can be very convenient to let that stay implicit so that you don't have to explicitly cast when passing a type `T` to a function accepting a `U`. 

Also bear in mind that's an API break. Not sure if we care about such an API break but we ought to know that users may be depending on the implicitness of the conversions at the moment.

File: moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp (hunk 3 above)
Comment (CursedRock17) [reply moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp]: Conflict Markers are resolved

File: moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h (hunk 1 above)
Comment (CursedRock17) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: Ok that makes sense, I was just trying to follow the cpp_best_practices in improving the codebase on general standards. Are there scenarios that we want to precisely target, if we don't want an API breakage in conversion operators, but still want implicit conversions at compile time? I don't know if there are defined situations in moveit 2 where it's needed on hand, outside from a search through comments explaining functions.

Comment (ChrisThrasher) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: > Are there scenarios that we want to precisely target, if we don't want an API breakage in conversion operators, but still want implicit conversions at compile time?

Can you explain what this means?

Comment (CursedRock17) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: I assumed when you said
> I agree that explicit is generally better than the lack of it but I'm 100% on board with explicit type conversion operators

You meant that there are some scenarios in this repository where `explicit` is useful to limit implicit conversions from happening and other instances when leaving it as implicit, so don't have to specify, is much easier, essentially it all depends on each use case for the function. I then, wanted to know if I needed to look, in depth, on each single parameter function that could utilize explicit and see if it is worth it to restrict that function. I apologize for miswording.

File: moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 4]
```diff
@@ -1,216 +0,0 @@
-/*********************************************************************
```
Comment (sjahr) [inline moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:1]: Pretty sure that you shouldn't delete this file, is this an accident?

Comment (sjahr) [reply moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:1]: Can you format the code correctly with pre-commit? You find instructions on how to do this here: https://moveit.ros.org/documentation/contributing/code/#pre-commit-formatting-checks

Comment (CursedRock17) [reply moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:1]: I must've accidentally deleted both moveit_cpp .h and .cpp, because I know I accidentally deleted the .cpp version and fixed that, so sorry about that. As for pre-commit I have ran it multiple times and it doesn't show errors, but I have cleaned up specifically where the CI was showcasing format-errors.

Comment (sjahr) [reply moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:1]: :point_up: @CursedRock17 I guess you need to push these changes

File: moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h (hunk 1 above)
Comment (ChrisThrasher) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: I would recommend against adding `explicit` to these conversion operators.

Comment (CursedRock17) [reply moveit_core/robot_model/include/moveit/robot_model/joint_model_group.h:79]: I have since, removed the explicit from all the operators that I have changed, but what about the many single parameter constructors? Most of then I have seen don't have a lot of of changes to be made within the codebase, but they're public so they would also technically break the API with many users.

File: moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 5]
```diff
@@ -175,8 +175,21 @@ class MoveItCpp
...
+  [[deprecated("MoveItCpp::execute() no longer requires a blocking parameter")]] moveit_controller_manager::ExecutionStatus
+  execute(const std::string& group_name, const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
+          bool blocking = true, const std::vector<std::string>& controllers = std::vector<std::string>());
+
+  moveit_controller_manager::ExecutionStatus
+<<<<<<< HEAD
+  execute(const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
+=======
+  execute(const robot_trajectory::RobotTrajectoryPtr& robot_trajectory, bool blocking = true,
+>>>>>>> parent of f117b24a6 (Removed Blocking)
```
Comment (sjahr) [inline moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:191]: This seems to be an artifact from rebasing

File: moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp [hunk 6]
```diff
@@ -245,6 +245,10 @@ MoveItCpp::execute(const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
   // Execute trajectory
   moveit_msgs::msg::RobotTrajectory robot_trajectory_msg;
   robot_trajectory->getRobotTrajectoryMsg(robot_trajectory_msg);
+  trajectory_execution_manager_->push(robot_trajectory_msg, controllers);
+  trajectory_execution_manager_->execute();
```
Comment (sjahr) [inline moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp:249]: These changes are not related to this PR

File: moveit_core/utils/include/moveit/utils/moveit_error_code.h [hunk 7]
```diff
@@ -55,7 +55,7 @@ class MoveItErrorCode : public moveit_msgs::msg::MoveItErrorCodes
   {
     val = code.val;
   }
-  explicit operator bool() const
```
Comment (sjahr) [inline moveit_core/utils/include/moveit/utils/moveit_error_code.h:58]: Why remove this here?

**Review by sjahr — CHANGES_REQUESTED** [review]
There seem to be artifacts that don't belong into this PR but are leftovers from a rebase

**Review by sjahr — CHANGES_REQUESTED** [review]
There seem to be artifacts that don't belong into this PR but are leftovers from a rebase

**Review by sjahr — CHANGES_REQUESTED** [review]
There seem to be artifacts that don't belong into this PR but are leftovers from a rebase

File: moveit_core/utils/include/moveit/utils/moveit_error_code.h (hunk 7 above)
Comment (CursedRock17) [reply moveit_core/utils/include/moveit/utils/moveit_error_code.h:58]: We had agreed earlier in the pull request to proceed with caution when it came to the explicit operator on type converters. If this example is safe, I can revert that change if that's what you expected here.

File: moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp (hunk 6 above)
Comment (CursedRock17) [reply moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp:249]: These were part of the function before the `blocking` parameter was removed [line 226](https://github.com/ros-planning/moveit2/blob/main/moveit_ros/planning/moveit_cpp/src/moveit_cpp.cpp) I don't know why it shows these are being new changes.

File: moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h [hunk 8]
```diff
@@ -175,6 +175,10 @@ class MoveItCpp
   execute(const std::string& group_name, const robot_trajectory::RobotTrajectoryPtr& robot_trajectory,
           bool blocking = true, const std::vector<std::string>& controllers = std::vector<std::string>());
 
+  [[deprecated("MoveItCpp::execute() no longer requires a blocking parameter")]] moveit_controller_manager::ExecutionStatus
```
Comment (sjahr) [inline moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:178]: Your rebase was not performed correctly because this is removed in the main branch. Please clean this up, you should only see the diffs you created here.

**Review by sjahr — CHANGES_REQUESTED** [review]

File: moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h (hunk 8 above)
Comment (CursedRock17) [reply moveit_ros/planning/moveit_cpp/include/moveit/moveit_cpp/moveit_cpp.h:178]: Oh I see, it was swapped with the last pull request I made in this repo. It'll be fixed soon, I apologize for the delay.

**Comment by tylerjw** [comment]
Any update on this PR?

**Comment by CursedRock17** [comment]
Honestly, no I haven't worked on this PR, since we agreed to leave the explicit operators alone I had worked on making `explicit` single parameter constructors. I thought I covered a decent amount of them, but I don't have the adequate knowledge of moveit2's codebase to know which ones will break, which are needed, or even where they all are. I don't know if I'm the correct person to implement the entirety of this PR.
